Rewrite chroot_setup in Python

markdirish opened the pull request:
Initial rough attempt at converting chroot_setup.sh to chroot_setup.py
Currently has a bugs, may not run, but every function save one has been converted to rough Python

Signed-off-by: Mark Irish <mirish@ibm.com>

markdirish commented:
Created draft pull request so @abmusse , @kadler , and myself have a place to discuss the code.

Code hasn't be intensively tested, I'm a bit of a coward with firing it off with all the `os` module calls inside of it. Pushed so that Abdirahim can see what I've done, discuss and help fix problems, and take over the effort himself if that's what he wants to do.

markdirish commented:
One thing I didn't know: In the `.sh` file, do  `chroot_ln_fix_rel` replace `chroot_ln_rel`? They appear to be doing similar things to my admittedly non-Bash eyes, so I only implemented the former.

abmusse commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+# Need to be done before any 'cd'
+TEMP_DIR = os.readlink(args.chroot_directory)
+
+logger.debug("Scipt Dir: " + SCRIPT_DIR)
```
The [default log level](https://docs.python.org/3.6/howto/logging.html#basic-logging-tutorial) is set to `WARNING` therefore all the logger.debug calls won't print unless the log level is adjusted. We could check if `-v` verbose mode was set and adjust the log level to something else i.e. `INFO` or `DEBUG`

```python
import logging
logging.warning('Watch out!')  # will print a message to the console
logging.info('I told you so')  # will not print anythinglogging.warning('Watch out!')  # will print a message to the console
logging.info('I told you so again!')  # will not print anything
```

abmusse commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+
+To set up your PATH to pick up RPM packages once inside your chroot
+Run: . /QOpenSys/etc/profile
+""")
```
Missing trailing new line

abmusse commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+    if ("ops" in arg):
+
+      # verify a deprecated OPS file was not specified
+      print("OPS has been depricated, and replaced with YUM Packages. Run $0 with a valid [CHROOT TYPE]")
```
We no longer even have the OPS lsts in our `config` dir so checking this check is probably obsolete and could be removed.

kadler replied:
Agreed, remove the check

abmusse commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+
+# TODO: documentation
+def initial_check ():
+  if (os.path.exists('/QOpenSys/usr/bin')):
```
The old script would check if the `/QOpenSys/usr/bin` path exists to determine if we are on IBM i. We could probably check `os.uname().sysname` now that we are using python.

kadler commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+# TODO: documentation
+def initial_check ():
+  if (os.path.exists('/QOpenSys/usr/bin')):
+    CHROOT_DEBUG = False
```
By Python conventions this would be a constant, but it's being altered below. Instead, we can drop this variable and reference args.verbose instead.

In addition, @abmusse mentioned using args.verbose to adjust the logging level and then just calling the appropriate logging function and we never have to even check the variable anywhere else.

kadler replied:
I guess this _is_ separate from verbosity of logging as when this is set, no action will be performed but it will still go through showing what would be done. Many utilities have command line options for this, usually with `--dry-run` or `-n`. I think that would make more sense to do something like that.

abmusse replied:
To me the `CHROOT_DEBUG` variable was a misnomer because it really used to guard against running commands on non IBM i sytems as it was only set to true in the else leg of `initial_check` when the script was run on non IBM i system then the script exited. 

Therefore none of the actions would run anyway so why did the script even bother to include checks for `CHROOT_DEBUG` in the actions.

abmusse replied:
I think we can get rid of all the `if not CHROOT_DEBUG` in the chroot_* actions. We just need to exit the script properly in the inital_check if we are not running on IBM i.

abmusse replied:
Fixed up in 1991e26

kadler commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+def initial_check ():
+  if (os.path.exists('/QOpenSys/usr/bin')):
+    CHROOT_DEBUG = False
+    PATH='/QOpenSys/usr/bin:/QOpenSys/usr/sbin:'
+    LIBPATH='/QOpenSys/usr/lib'
```
These variables could be gotten rid of. Set os.environ and then reference it later if needed.

abmusse replied:
We no longer set PATH or LIBPATH See e6ff3cf and 7a5e93a

kadler commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+
+#
+# MAIN
+#
```
Mainline code should be under an `if __name__ == '__main__'` block

abmusse replied:
Moved to main block in 10a7597

kadler commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+# perform installs if set
+yum_install(args.yum_installs)
+
+print("""
+
+Done creating your chroot!
+
+To enter your chroot
+Run: chroot """ + args.chroot_directory + """ /QOpenSys/usr/bin/sh
```
```suggestion
print(f"""

Done creating your chroot!

To enter your chroot
Run: chroot {args.chroot_directory} /QOpenSys/usr/bin/sh
```

abmusse replied:
This looks resolved.

kadler commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+    if not val == "y":
+      print("Bye")
+      sys.exit(-4)
+  else:
+    # continues automatically
```
```suggestion
```

abmusse replied:
This looks resolved in https://github.com/IBM/ibmichroot/pull/79/commits/57be2ef025bcd7daa9428281106ba86e167ebaa1

kadler commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+  os.mkdir(CHROOT_DIR)
+  # TODO: Check creation
+else:
+  if not args.auto_yes
+    print(CHROOT_DIR + " directory already exists")
+    val = input("Would you like to continue chroot setup? [y/N]")
```
We could probably do something like this to make this more DRY:

```python

def input_yes(*args, **kwargs):
    return 'y'

if args.auto_yes:
    input = input_yes
```

abmusse replied:
Fixed up in d80220a

kadler reviewed (commented):
Preliminary comments. This seems like a straight-forward port from the original shell script. While that definitely has improvements over the shell script, I'd like to go even further.

markdirish commented:
Sorry, didn't see that @abmusse unassigned himself from #70. Will take a stab at this once I finish up some of the rpm issues I am working on (and stabilize some `sequelize-ibmi` and `odbc` issues).

markdirish commented:
Ok I changed the name and apparently it didn't like that for keeping track of everything.

What I changed:

* Made sure that all functions are passed a single string that represents the line from the file. Then each function has to call line.split() to split it into tokens, which it then parses.
* Create `assert` statements in functions to make sure that the right number of tokens are passed for each action
* Print out every Python command that will be run, instead of the terminal command. (These should probably be hidden behind the `-v` flag I would think, most people probably don't care about all of that output).
* Rewrite some functions like `chroot_mknod` to make sure that it parses to get what the Python functions are expecting.

I have run this with DEBUG = True, so that it tells me what it will run but doesn't actually run it. I think I will try to run these functions/actions piecemeal, as I'm afraid of them doing something bad like mucking with the entire file structure. I suppose I could try to run it inside of a chroot (to create a sub chroot?), and that way I wouldn't have to be afraid...

markdirish commented:
Ok! This is finally to a stage where it is working and I feel confident that it can do what it needs to do. I built with a `chroot_minimal.lst` and it (mostly) worked. I say mostly because Python chokes on things that running the bash utilities doesn't. Thinks like trying to symlink things that already exist, or copying files/directories that don't exist. I've made some changes to that `.lst` to get it to a point where this program doesn't choke.

The only thing now is to clean it up, change wording on things like help text (or that goofy ASCII text art), and refactor where desired.

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+        sys.exit(-5)
+
+    mylist = []
+    mylist.append(SCRIPT_DIR + "/chroot_includes.lst")
+    mylist.append(SCRIPT_DIR + "/chroot_minimal.lst")
```
After #75, we should now use default.lst instead.

```suggestion
mylist.append(SCRIPT_DIR + "/default.lst")
```

abmusse replied:
I will push a commit with this change

abmusse replied:
This branch would need to be rebased with master to pull in default.lst, for testing purposes.

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+    # Only chroot_directory was provided to the script:
+    # Ask if minimal with includes chroot is desired
+    if not auto_yes:
+      val = input("Would you like to create a minimal chroot w/ includes @ " + CHROOT_DIR + "? [y/N]:")
```
```suggestion
      val = input("Would you like to create a default chroot environment @ " + CHROOT_DIR + "? [y/N]:")
```

abmusse replied:
I will push a commit with this change

kadler commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+def chroot_ln (line: str):
+  ln_command = line.split()
+  assert len(ln_command) == 2, ":ln action expects 2 arguments!"
+  target = ln_command[0]
+  link_name = ln_command[1]
```
This can be done a bit more succinctly:

```suggestion
  target, link_name = ln_command
```

abmusse replied:
Fixed up in 04ca7a33a71c868370b6315b537307b3984e03c2

kadler commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+# Python won't check and therefore won't throw an error.
+def chroot_ln (line: str):
+  ln_command = line.split()
+  assert len(ln_command) == 2, ":ln action expects 2 arguments!"
```
This should probably raise a ValueError instead of asserting

abmusse replied:
Similar asserts are done in the other `chroot_*` functions

abmusse replied:
Fixed up in b12fbbc

kadler commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+    print("setting up based on " + chroot_file)
+    print("=====================================")
+
+    for key in globals_dictionary:
+      os.environ[key] = globals_dictionary[key]
+
+    chroot_setup(chroot_file)
+
+    for key in globals_dictionary:
+      del os.environ[key]
```
Personally, I think we should stop using environment variables for this behavior and use a dictionary for the passed mappings instead. Of course, this does break any scripts that relied on environment variables which were set prior to invoking the script. If such list files exist, we could add a fallback to check the environ if it's not found in the mapping.

abmusse replied:
Actually these globals were never used as enviornmnet variables in the original script. These are named variables passed into chroot_setup.sh that get  substitute placeholder values in the xxx.lst files. See https://github.com/IBM/ibmichroot#advanced.

abmusse replied:
Fixed up in ffb07ef

kadler commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+
+  os.makedirs(args.chroot_directory + '/QOpenSys/etc', exist_ok=True)
+  etc_profile = (args.chroot_directory + '/QOpenSys/etc/profile')
+  if not os.path.isfile(etc_profile):
+    print('Priming /QOpenSys/etc/profile')
+    etc_file = open(etc_profile, 'a')
```
We check if the file doesn't exist, but then open it for appending? If we want to ensure that the file is only written to if doesn't already exist, we should use the 'x' flag:

```python
with open(etc_profile, 'wx') as etc_file:
    print('Priming /QOpenSys/etc/profile')
    ...
```

abmusse replied:
Fixed up in f8c5350b141afd53c324da0d4ccf2a3303661f08

kadler commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+  if not os.path.isfile(etc_profile):
+    print('Priming /QOpenSys/etc/profile')
+    etc_file = open(etc_profile, 'a')
+    etc_file.write('PATH=/QOpenSys/pkgs/bin:$PATH\n')
```
FYI, you can use print for this to add the newline

```suggestion
    print('PATH=/QOpenSys/pkgs/bin:$PATH', file=etc_file)
```

abmusse replied:
fixed up in f8c5350b141afd53c324da0d4ccf2a3303661f08

kadler commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+      if not CHROOT_LIST.endswith('.lst'):
+        CHROOT_LIST = CHROOT_LIST + '.lst'
+
+      # check that the .lst file exists
+      if not os.path.isfile(CHROOT_LIST):
+        print(f"{CHROOT_LIST} cannot be found in {os.cwd}")
+        sys.exit(-7)
+      else:
+        # append file name to end of list
+        mylist.append(os.readlink(CHROOT_LIST))
```
Probably should open them too, while we're at it. If we still want to print the file path later, we could store a tuple(path, file obj) in the array.

abmusse replied:
Fixed up in d80220a

Instead of checking if the file exists I just append the file name to the last. later in the chroot_setup function the file is opened for reading.

kadler commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+      continue
+
+    # Determine action
+    elif line.startswith(':'):
+      action = action_dictionary[line]
+
+    # Execute action
+    else:
+      if action is not None:
+        action(line)
+
+# Mapping the .lst action keys to the functions that run on their values
+action_dictionary = {
+  ':file'       : chroot_setup,
+  ':mkdir'      : chroot_mkdir,      # done
+  ':ln_fix_rel' : chroot_ln_fix_rel, # done
+  ':ln_rel'     : chroot_ln_rel,     # done
+  ':ln'         : chroot_ln,         # done
+  ':mknod'      : chroot_mknod,      # done
+  ':cp_dir'     : chroot_cp_dir,     # done
+  ':cp'         : chroot_cp,         # done
+  ':chmod_dir'  : chroot_chmod_dir,  # done
+  ':chmod'      : chroot_chmod,      # done
+  ':chown_dir'  : chroot_chown_dir,  # done
+  ':chown'      : chroot_chown,      # done
+  ':system'     : chroot_system,     # done
+  ':sh'         : chroot_sh          # done
+}
```
If we wanted to enforce a consistent function naming scheme, this mapping could go away:

```python
    elif line.startswith(':'):
        function = f"chroot_{line[1:]}"
        action = globals()[function]
```

In addition, it might be good to catch the KeyError if the user gives an unexpected action.

abmusse replied:
The only inconsistent action -> function name is:

```
':file'       : chroot_setup,
```

We could make a wrapper `chroot_file` function that calls `chroot_setup`

abmusse replied:
Fixed up in 
ca5966b

kadler commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+    PATH='/QOpenSys/pkgs/bin:/QOpenSys/pkgs/sbin:/QOpenSys/usr/bin:/QOpenSys/usr/sbin:'
+    LIBPATH='/QOpenSys/pkgs/lib:/QOpenSys/usr/lib'
+    os.environ['PATH'] = PATH
+    os.environ['LIBPATH'] = LIBPATH
```
We really shouldn't be setting LIBPATH.

abmusse replied:
Fixed this up in 7a5e93adb7b1ac5f3bee0756ca240518a945f4c0

kadler commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+def initial_check():
+  if (os.uname().sysname == 'OS400'):
+    # CHROOT_DEBUG = False
+    PATH='/QOpenSys/pkgs/bin:/QOpenSys/pkgs/sbin:/QOpenSys/usr/bin:/QOpenSys/usr/sbin:'
```
Not sure why we're setting PATH. It's also odd that we're leaving the last element empty. This has the same effect as "." - adding the current directory to the PATH, which is arguably not a good idea.

abmusse replied:
Fixed this up in e6ff3cfda15c3ceb37e83c23e1bbae7f0e9dbbe0

kadler reviewed (commented):
In addition to the specific comments, I'd also like to add that the code should be run through pylint - for sure the indentation should be adjusted to 4 spaces.

abmusse commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+def chroot_mkdir ( dir: str ):
+  # TODO: fix print statement
+  print('os.mkdirs(' + args.chroot_directory + dir + ', exist_ok=True)')
+  if (CHROOT_DEBUG == False):
```
Fixed this up in 177c0a58c5f27478a547732d683d5a203b3ada97

abmusse commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+# TODO: documentation
+def chroot_mkdir ( dir: str ):
+  # TODO: fix print statement
+  print('os.mkdirs(' + args.chroot_directory + dir + ', exist_ok=True)')
```
fstrings are now used here but still doing concat with `+` operator: 
`print(f'os.makedirs({args.chroot_directory + dir}, exist_ok=True)')`

Will fix these up to use f-strings

abmusse replied:
Fixed these up in e0a773d

abmusse commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+
+    # Comment
+    else if line.startswith('#'):
+      # print('comment')
```
This looks resolved

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+parser.add_argument("-y", dest='auto_yes', action='store_true', help="Automatically answer yes to input prompts")
+parser.add_argument("-v", dest='verbose', action='store_true', help="Print verbose output when running")
+args = parser.parse_args()
+parser.parse_args()
```
Fixed this up in 10a7597

abmusse commented on chroot_setup.py:
```diff
@@ -0,0 +1,373 @@
+
+    # Determine action
+    else if line.startswith(':'):
+      action = action_dictionary[line]
```
Fixed this up in 1c33b97

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,434 @@
+    if (args.verbose):
+      print(f"PATH={PATH}")
+
+  else:
+    # CHROOT_DEBUG = True
+    print("**********************")
+    print("Not IBM i, no action is taken (debug flow purpose only).")
+    print("**********************")
+    # exit -1
```
This is the else leg that indicates we are not running chroot setup on IBM i therefor we should exit after printing the error message.

abmusse replied:
Fixed this up in dd947c5

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,434 @@
+  #####  #######    #     #####  #
+""")
+
+CHROOT_DEBUG : bool = False
```
The value of CHROOT_DEBUG is never changed

abmusse replied:
Fixed up in 1991e26

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,430 @@
+  args = parser.parse_args()
+  parser.parse_args()
+
+  lst = 0
+  ops = 0
+  qsys = 0
```
These three variables are unused.

abmusse replied:
Fixed this up in 05412ed

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+    # directory that exists at that location.
+    if os.path.exists(args.chroot_directory + line):
+      shutil.rmtree(args.chroot_directory + line)
+    shutil.copytree(line, args.chroot_directory + line)
```
Added this check in 22b0b1d6b21bb138bf2b466d2b310d6ddb3e4062

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,432 @@
+  line = line.rstrip()
+  print(f"system -i {line}")
+  if not CHROOT_DEBUG:
+    os.system(f"system -i {line}")
```
We should use [subprocess.run](https://docs.python.org/3/library/subprocess.html?ref=hackernoon.com#subprocess.run) instead of `os.system`

abmusse replied:
Fixed this up in 3814d9b

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,432 @@
+def chroot_sh(line: str):
+  print(line)
+  if not CHROOT_DEBUG:
+    os.system(line)
```
See https://github.com/IBM/ibmichroot/pull/79/files#r565660065

abmusse replied:
Fixed this up in 3814d9b

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,432 @@
+  for package in package_list:
+    command = install_template.substitute(yes=auto_yes, chroot_dir=args.chroot_directory, package=package)
+    logging.debug(command)
+    os.system(command)
```
dito

abmusse replied:
Fixed this up in  3814d9b

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,440 @@
+  elif (mknod_command[1] == 'b'):
+    mode = 0o644 | stat.S_IFBLK
+  elif (mknod_command[1] == 'c'):
+    mode = 0o644 | stat.S_IFCHR
```
Fixed this up in 3464874400d88f3e5f5d4e55688999c1d3013fe5

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,428 @@
+  parser = argparse.ArgumentParser()
+  # positional arguments
+  parser.add_argument("chroot_directory", metavar="CHROOT_DIRECTORY", help="The location of the chroot to be created")
+  parser.add_argument("chroot_type", metavar="CHROOT_TYPE", help="A list of .lst files to use for generating the chroot", nargs="?")
```
chroot_type `nargs` should `*` (zero or more) instead of `?` (0 or 1)  because we can specify multiple lst files e.g. `minimal`,  `includes`, `nls`

abmusse replied:
Fixed this up in d80220a

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,428 @@
+  print(f"os.remove({args.chroot_directory}{node})")
+  print(f"os.mknod({args.chroot_directory}{node}, {mode}, {device})")
+  if not CHROOT_DEBUG:
+    os.remove(f"{args.chroot_directory}{node}")
```
This remove should be wrapped in a try catch. It should only be removed when the it already exists. On initial chroot setup it does not exist and this line raises a `FileNotFoundError`.

```suggestion
    try:
      os.remove(f"{args.chroot_directory}{node}")
    except FileNotFoundError:
      pass
```

abmusse replied:
Fixed up in e1164d1

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,435 @@
+
+  mode = int(chmod_command[0], 8)
+  chmod_file = f"{args.chroot_directory}{chmod_command[1]}"
+  printf(f"os.chmod({chmod_file}, {mode})")
```
Typo should be print

abmusse replied:
fixed up in 6022fb9

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,435 @@
+
+  user = chown_command[0]
+  chown_file = f"{args.chroot_directory}{chown_command[1]}"
+  printf(f"os.chown({args.chroot_directory}{chown_file}, {user})")
```
dito https://github.com/IBM/ibmichroot/pull/79/files#r574193939

abmusse replied:
fixed up in 6022fb9

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,435 @@
+  chown_directory = f"{args.chroot_directory}{chown_command[1]}"
+  for root, dirs, files in os.walk(f"{args.chroot_directory}{chown_directory}"):
+    for dir in dirs:
+      printf(f"os.chown({os.path.join(root, dir)}, {user})")
```
dito https://github.com/IBM/ibmichroot/pull/79/files#r574193939

abmusse replied:
fixed up in 6022fb9

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,435 @@
+      printf(f"os.chown({os.path.join(root, dir)}, {user})")
+      os.chown(os.path.join(root, dir), user)
+    for file in files:
+      printf(f"os.chown({os.path.join(root, file)}, {user})")
```
dito https://github.com/IBM/ibmichroot/pull/79/files#r574193939

abmusse replied:
fixed up in 6022fb9

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,435 @@
+  here = os.getcwd()
+  print(f"os.chdir({args.chroot_directory}{directory})")
+  print(f"os.symlink({target}, {basename})")
+  os.chdir(f"{args.chroot_directory}{mydir}")
```
mydir is undefined

abmusse replied:
fixed up in 6022fb9

abmusse commented on chroot_setup:
```diff
@@ -0,0 +1,435 @@
+  print(f"os.chdir({args.chroot_directory}{directory})")
+  print(f"os.symlink{target}, {basename})")
+  print(f"os.chdir({here})")
+  os.chdir(f"{args.chroot_directory}{mydir}")
```
dito https://github.com/IBM/ibmichroot/pull/79/files#r574194818

abmusse replied:
fixed up in 6022fb9

abmusse commented:
@kadler I've fixed up the requested changes and added flake8 linter action
 derived from default [pythonpackage.yml](https://docs.github.com/en/actions/guides/building-and-testing-python)

After fixing up initial linter issues `undefined names` in 6022fb9.

We now have a bunch of cleaning up todo:

```
chroot_setup:12:1: F401 'enum.Enum' imported but unused
chroot_setup:16:46: W291 trailing whitespace
chroot_setup:17:46: W291 trailing whitespace
chroot_setup:18:46: W291 trailing whitespace
chroot_setup:19:46: W291 trailing whitespace
chroot_setup:20:46: W291 trailing whitespace
chroot_setup:21:46: W291 trailing whitespace
chroot_setup:22:1: W293 blank line contains whitespace
chroot_setup:23:40: W291 trailing whitespace
chroot_setup:24:41: W291 trailing whitespace
chroot_setup:25:41: W291 trailing whitespace
chroot_setup:26:40: W291 trailing whitespace
chroot_setup:27:35: W291 trailing whitespace
chroot_setup:28:35: W291 trailing whitespace
chroot_setup:34:1: E302 expected 2 blank lines, found 1
chroot_setup:34:17: E211 whitespace before '('
chroot_setup:35:3: E111 indentation is not a multiple of four
chroot_setup:36:3: E111 indentation is not a multiple of four
chroot_setup:42:1: E302 expected 2 blank lines, found 1
chroot_setup:42:14: E211 whitespace before '('
chroot_setup:43:3: E111 indentation is not a multiple of four
chroot_setup:45:3: E111 indentation is not a multiple of four
chroot_setup:48:3: E111 indentation is not a multiple of four
chroot_setup:50:3: E111 indentation is not a multiple of four
chroot_setup:51:3: E111 indentation is not a multiple of four
chroot_setup:56:1: E302 expected 2 blank lines, found 1
chroot_setup:56:18: E211 whitespace before '('
chroot_setup:57:3: E111 indentation is not a multiple of four
chroot_setup:59:3: E111 indentation is not a multiple of four
chroot_setup:62:3: E111 indentation is not a multiple of four
chroot_setup:63:3: E111 indentation is not a multiple of four
chroot_setup:64:3: E111 indentation is not a multiple of four
chroot_setup:65:3: E111 indentation is not a multiple of four
chroot_setup:66:3: E111 indentation is not a multiple of four
chroot_setup:67:3: E111 indentation is not a multiple of four
chroot_setup:68:3: E111 indentation is not a multiple of four
chroot_setup:69:3: E111 indentation is not a multiple of four
chroot_setup:70:3: E111 indentation is not a multiple of four
chroot_setup:76:1: E302 expected 2 blank lines, found 1
chroot_setup:76:22: E211 whitespace before '('
chroot_setup:77:3: E111 indentation is not a multiple of four
chroot_setup:79:3: E111 indentation is not a multiple of four
chroot_setup:82:3: E111 indentation is not a multiple of four
chroot_setup:83:3: E111 indentation is not a multiple of four
chroot_setup:84:3: E111 indentation is not a multiple of four
chroot_setup:85:3: E111 indentation is not a multiple of four
chroot_setup:86:3: E111 indentation is not a multiple of four
chroot_setup:87:3: E111 indentation is not a multiple of four
chroot_setup:88:3: E111 indentation is not a multiple of four
chroot_setup:89:3: E111 indentation is not a multiple of four
chroot_setup:90:3: E111 indentation is not a multiple of four
chroot_setup:91:3: E111 indentation is not a multiple of four
chroot_setup:95:1: E302 expected 2 blank lines, found 1
chroot_setup:95:17: E211 whitespace before '('
chroot_setup:96:3: E114 indentation is not a multiple of four (comment)
chroot_setup:97:3: E114 indentation is not a multiple of four (comment)
chroot_setup:98:3: E111 indentation is not a multiple of four
chroot_setup:100:3: E111 indentation is not a multiple of four
chroot_setup:103:3: E111 indentation is not a multiple of four
chroot_setup:103:12: E221 multiple spaces before operator
chroot_setup:105:3: E114 indentation is not a multiple of four (comment)
chroot_setup:106:3: E111 indentation is not a multiple of four
chroot_setup:110:3: E111 indentation is not a multiple of four
chroot_setup:112:3: E114 indentation is not a multiple of four (comment)
chroot_setup:113:3: E111 indentation is not a multiple of four
chroot_setup:114:3: E111 indentation is not a multiple of four
chroot_setup:115:3: E111 indentation is not a multiple of four
chroot_setup:117:3: E111 indentation is not a multiple of four
chroot_setup:118:3: E111 indentation is not a multiple of four
chroot_setup:119:3: E111 indentation is not a multiple of four
chroot_setup:121:3: E111 indentation is not a multiple of four
chroot_setup:124:3: E111 indentation is not a multiple of four
chroot_setup:129:1: E302 expected 2 blank lines, found 1
chroot_setup:129:14: E211 whitespace before '('
chroot_setup:130:3: E111 indentation is not a multiple of four
chroot_setup:131:3: E111 indentation is not a multiple of four
chroot_setup:136:1: E302 expected 2 blank lines, found 1
chroot_setup:136:18: E211 whitespace before '('
chroot_setup:137:3: E111 indentation is not a multiple of four
chroot_setup:138:3: E114 indentation is not a multiple of four (comment)
chroot_setup:139:3: E114 indentation is not a multiple of four (comment)
chroot_setup:140:3: E114 indentation is not a multiple of four (comment)
chroot_setup:141:3: E114 indentation is not a multiple of four (comment)
chroot_setup:142:3: E111 indentation is not a multiple of four
chroot_setup:142:28: E231 missing whitespace after ','
chroot_setup:144:3: E111 indentation is not a multiple of four
chroot_setup:146:7: E111 indentation is not a multiple of four
chroot_setup:151:1: E302 expected 2 blank lines, found 1
chroot_setup:151:17: E211 whitespace before '('
chroot_setup:152:3: E111 indentation is not a multiple of four
chroot_setup:154:3: E111 indentation is not a multiple of four
chroot_setup:157:3: E111 indentation is not a multiple of four
chroot_setup:158:3: E111 indentation is not a multiple of four
chroot_setup:159:3: E111 indentation is not a multiple of four
chroot_setup:160:3: E111 indentation is not a multiple of four
chroot_setup:164:1: E302 expected 2 blank lines, found 1
chroot_setup:164:21: E211 whitespace before '('
chroot_setup:165:3: E111 indentation is not a multiple of four
chroot_setup:167:3: E111 indentation is not a multiple of four
chroot_setup:170:3: E111 indentation is not a multiple of four
chroot_setup:171:3: E111 indentation is not a multiple of four
chroot_setup:172:3: E111 indentation is not a multiple of four
chroot_setup:174:7: E111 indentation is not a multiple of four
chroot_setup:175:7: E111 indentation is not a multiple of four
chroot_setup:177:7: E111 indentation is not a multiple of four
chroot_setup:178:7: E111 indentation is not a multiple of four
chroot_setup:182:1: E302 expected 2 blank lines, found 1
chroot_setup:182:17: E211 whitespace before '('
chroot_setup:183:3: E111 indentation is not a multiple of four
chroot_setup:185:3: E111 indentation is not a multiple of four
chroot_setup:188:3: E111 indentation is not a multiple of four
chroot_setup:189:3: E111 indentation is not a multiple of four
chroot_setup:190:3: E111 indentation is not a multiple of four
chroot_setup:191:3: E111 indentation is not a multiple of four
chroot_setup:195:1: E302 expected 2 blank lines, found 1
chroot_setup:195:21: E211 whitespace before '('
chroot_setup:196:3: E111 indentation is not a multiple of four
chroot_setup:198:3: E111 indentation is not a multiple of four
chroot_setup:201:3: E111 indentation is not a multiple of four
chroot_setup:202:3: E111 indentation is not a multiple of four
chroot_setup:203:3: E111 indentation is not a multiple of four
chroot_setup:205:7: E111 indentation is not a multiple of four
chroot_setup:206:7: E111 indentation is not a multiple of four
chroot_setup:208:7: E111 indentation is not a multiple of four
chroot_setup:209:7: E111 indentation is not a multiple of four
chroot_setup:213:1: E302 expected 2 blank lines, found 1
chroot_setup:214:3: E111 indentation is not a multiple of four
chroot_setup:215:3: E111 indentation is not a multiple of four
chroot_setup:216:3: E111 indentation is not a multiple of four
chroot_setup:221:1: E302 expected 2 blank lines, found 1
chroot_setup:222:3: E111 indentation is not a multiple of four
chroot_setup:223:3: E111 indentation is not a multiple of four
chroot_setup:224:3: E111 indentation is not a multiple of four
chroot_setup:228:1: E302 expected 2 blank lines, found 1
chroot_setup:229:3: E111 indentation is not a multiple of four
chroot_setup:231:1: E302 expected 2 blank lines, found 1
chroot_setup:231:17: E211 whitespace before '('
chroot_setup:232:3: E111 indentation is not a multiple of four
chroot_setup:236:7: E111 indentation is not a multiple of four
chroot_setup:238:11: E114 indentation is not a multiple of four (comment)
chroot_setup:239:11: E111 indentation is not a multiple of four
chroot_setup:241:7: E111 indentation is not a multiple of four
chroot_setup:243:7: E114 indentation is not a multiple of four (comment)
chroot_setup:244:7: E111 indentation is not a multiple of four
chroot_setup:247:7: E114 indentation is not a multiple of four (comment)
chroot_setup:248:7: E111 indentation is not a multiple of four
chroot_setup:249:11: E114 indentation is not a multiple of four (comment)
chroot_setup:250:11: E114 indentation is not a multiple of four (comment)
chroot_setup:251:11: E114 indentation is not a multiple of four (comment)
chroot_setup:252:11: E114 indentation is not a multiple of four (comment)
chroot_setup:253:11: E114 indentation is not a multiple of four (comment)
chroot_setup:254:11: E111 indentation is not a multiple of four
chroot_setup:255:11: E111 indentation is not a multiple of four
chroot_setup:258:11: E111 indentation is not a multiple of four
chroot_setup:262:7: E114 indentation is not a multiple of four (comment)
chroot_setup:263:7: E111 indentation is not a multiple of four
chroot_setup:265:11: E111 indentation is not a multiple of four
chroot_setup:269:1: E302 expected 2 blank lines, found 1
chroot_setup:270:3: E111 indentation is not a multiple of four
chroot_setup:275:3: E111 indentation is not a multiple of four
chroot_setup:284:3: E111 indentation is not a multiple of four
chroot_setup:285:3: E111 indentation is not a multiple of four
chroot_setup:286:3: E111 indentation is not a multiple of four
chroot_setup:287:3: E111 indentation is not a multiple of four
chroot_setup:288:3: E111 indentation is not a multiple of four
chroot_setup:289:3: E111 indentation is not a multiple of four
chroot_setup:291:1: E302 expected 2 blank lines, found 1
chroot_setup:292:3: E111 indentation is not a multiple of four
chroot_setup:297:1: E305 expected 2 blank lines after class or function definition, found 1
chroot_setup:297:1: C901 'If 297' is too complex (19)
chroot_setup:298:3: E111 indentation is not a multiple of four
chroot_setup:299:3: E114 indentation is not a multiple of four (comment)
chroot_setup:300:3: E111 indentation is not a multiple of four
chroot_setup:301:3: E111 indentation is not a multiple of four
chroot_setup:301:128: E501 line too long (132 > 127 characters)
chroot_setup:302:3: E114 indentation is not a multiple of four (comment)
chroot_setup:303:3: E111 indentation is not a multiple of four
chroot_setup:303:128: E501 line too long (128 > 127 characters)
chroot_setup:304:3: E111 indentation is not a multiple of four
chroot_setup:304:128: E501 line too long (131 > 127 characters)
chroot_setup:305:3: E111 indentation is not a multiple of four
chroot_setup:306:3: E111 indentation is not a multiple of four
chroot_setup:307:3: E111 indentation is not a multiple of four
chroot_setup:308:3: E111 indentation is not a multiple of four
chroot_setup:310:3: E111 indentation is not a multiple of four
chroot_setup:311:3: E111 indentation is not a multiple of four
chroot_setup:312:3: E111 indentation is not a multiple of four
chroot_setup:314:3: E111 indentation is not a multiple of four
chroot_setup:316:3: E111 indentation is not a multiple of four
chroot_setup:317:3: E111 indentation is not a multiple of four
chroot_setup:319:3: E114 indentation is not a multiple of four (comment)
chroot_setup:320:3: E114 indentation is not a multiple of four (comment)
chroot_setup:321:3: E111 indentation is not a multiple of four
chroot_setup:322:3: E111 indentation is not a multiple of four
chroot_setup:324:3: E114 indentation is not a multiple of four (comment)
chroot_setup:325:3: E114 indentation is not a multiple of four (comment)
chroot_setup:326:3: E111 indentation is not a multiple of four
chroot_setup:329:3: E111 indentation is not a multiple of four
chroot_setup:330:3: E111 indentation is not a multiple of four
chroot_setup:331:3: E111 indentation is not a multiple of four
chroot_setup:333:3: E111 indentation is not a multiple of four
chroot_setup:338:3: E111 indentation is not a multiple of four
chroot_setup:340:3: E111 indentation is not a multiple of four
chroot_setup:344:3: E111 indentation is not a multiple of four
chroot_setup:346:7: E111 indentation is not a multiple of four
chroot_setup:352:7: E111 indentation is not a multiple of four
chroot_setup:353:7: E111 indentation is not a multiple of four
chroot_setup:356:3: E303 too many blank lines (2)
chroot_setup:356:3: E114 indentation is not a multiple of four (comment)
chroot_setup:357:3: E111 indentation is not a multiple of four
chroot_setup:359:3: E111 indentation is not a multiple of four
chroot_setup:364:7: E111 indentation is not a multiple of four
chroot_setup:368:7: E111 indentation is not a multiple of four
chroot_setup:369:7: E111 indentation is not a multiple of four
chroot_setup:370:7: E111 indentation is not a multiple of four
chroot_setup:375:3: E111 indentation is not a multiple of four
chroot_setup:378:7: E111 indentation is not a multiple of four
chroot_setup:380:7: E111 indentation is not a multiple of four
chroot_setup:382:7: E114 indentation is not a multiple of four (comment)
chroot_setup:383:7: E111 indentation is not a multiple of four
chroot_setup:386:7: E114 indentation is not a multiple of four (comment)
chroot_setup:387:7: E111 indentation is not a multiple of four
chroot_setup:390:7: E114 indentation is not a multiple of four (comment)
chroot_setup:391:7: E111 indentation is not a multiple of four
chroot_setup:393:3: E111 indentation is not a multiple of four
chroot_setup:394:3: E111 indentation is not a multiple of four
chroot_setup:396:3: E114 indentation is not a multiple of four (comment)
chroot_setup:397:3: E114 indentation is not a multiple of four (comment)
chroot_setup:398:3: E111 indentation is not a multiple of four
chroot_setup:399:3: E111 indentation is not a multiple of four
chroot_setup:400:7: E111 indentation is not a multiple of four
chroot_setup:404:3: E111 indentation is not a multiple of four
chroot_setup:410:3: E111 indentation is not a multiple of four
chroot_setup:411:3: E111 indentation is not a multiple of four
chroot_setup:412:3: E111 indentation is not a multiple of four
chroot_setup:414:7: E111 indentation is not a multiple of four
chroot_setup:415:7: E111 indentation is not a multiple of four
chroot_setup:416:7: E111 indentation is not a multiple of four
chroot_setup:417:7: E111 indentation is not a multiple of four
chroot_setup:418:7: E111 indentation is not a multiple of four
chroot_setup:419:3: E111 indentation is not a multiple of four
chroot_setup:422:3: E114 indentation is not a multiple of four (comment)
chroot_setup:423:3: E111 indentation is not a multiple of four
chroot_setup:426:3: E111 indentation is not a multiple of four
1     C901 'If 297' is too complex (19)
163   E111 indentation is not a multiple of four
30    E114 indentation is not a multiple of four (comment)
12    E211 whitespace before '('
1     E221 multiple spaces before operator
1     E231 missing whitespace after ','
17    E302 expected 2 blank lines, found 1
1     E303 too many blank lines (2)
1     E305 expected 2 blank lines after class or function definition, found 1
3     E501 line too long (132 > 127 characters)
1     F401 'enum.Enum' imported but unused
12    W291 trailing whitespace
1     W293 blank line contains whitespace
244
```

Note these linter issues are considered warnings.
the action looks for syntax errors or undefined names and considers those fatal.

```yml
...

      run: |
        # stop the build if there are Python syntax errors or undefined names
        flake8 chroot_setup --count --select=E9,F63,F7,F82 --show-source --statistics
        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
        flake8 chroot_setup --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics

...
```
